sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala

```diff
@@ -166,26 +166,63 @@ abstract class Star extends LeafExpression with NamedExpression {
  * "SELECT * FROM ...".
  *
- * @param table an optional table that should be the target of the expansion.  If omitted all
- *              tables' columns are produced.
+ * This is also used to expand structs. For example:
+ * "SELECT record.* from (SELECT struct(a,b,c) as record ...)
+ *
+ * @param target an optional name that should be the target of the expansion.  If omitted all
+ *              targets' columns are produced. This can either be a table name or struct name. This
+ *              is a list of identifiers that is the path of the expansion.
  */
-case class UnresolvedStar(table: Option[String]) extends Star with Unevaluable {
+case class UnresolvedStar(target: Option[Seq[String]]) extends Star with Unevaluable {
+
+  override def expand(input: LogicalPlan, resolver: Resolver): Seq[NamedExpression] = {
 
-  override def expand(input: Seq[Attribute], resolver: Resolver): Seq[NamedExpression] = {
-    val expandedAttributes: Seq[Attribute] = table match {
+    // First try to expand assuming it is table.*.
+    val expandedAttributes: Seq[Attribute] = target match {
       // If there is no table specified, use all input attributes.
-      case None => input
+      case None => input.output
       // If there is a table, pick out attributes that are part of this table.
-      case Some(t) => input.filter(_.qualifiers.filter(resolver(_, t)).nonEmpty)
+      case Some(t) => if (t.size == 1) {
+        input.output.filter(_.qualifiers.filter(resolver(_, t.head)).nonEmpty)
+      } else {
+        List()
+      }
     }
-    expandedAttributes.zip(input).map {
+    expandedAttributes.zip(input.output).map {
       case (n: NamedExpression, _) => n
       case (e, originalAttribute) =>
         Alias(e, originalAttribute.name)(qualifiers = originalAttribute.qualifiers)
     }
+
+    if (!expandedAttributes.isEmpty) return expandedAttributes
+
+    // Try to resolve it as a struct expansion. If there is a conflict and both are possible,
+    // (i.e. [name].* is both a table and a struct), the struct path can always be qualified.
+    if (target.isDefined) {
+      val attribute = input.resolve(target.get, resolver)
+      if (attribute.isDefined) {
+        // This target resolved to an attribute in child. It must be a struct. Expand it.
+        attribute.get.dataType match {
+          case s: StructType => {
+            s.fields.map( f => {
+              val extract = GetStructField(attribute.get, f, s.getFieldIndex(f.name).get)
+              Alias(extract, target.get + "." + f.name)()
+            })
+          }
+          case _ => {
+            throw new AnalysisException("Can only star expand struct data types. Attribute: `" +
+              target.get + "`")
+          }
+        }
+      } else {
+        List()
+      }
+    } else {
+      List()
+    }
   }
 
-  override def toString: String = table.map(_ + ".").getOrElse("") + "*"
+  override def toString: String = target.map(_ + ".").getOrElse("") + "*"
```

Comment:
`map(_.mkString("."))` instead of `map(_ + ".")`?

```
scala> Some(Seq("str1", "str2"))
res0: Some[Seq[String]] = Some(List(str1, str2))

scala> res0.map(_ + ".")
res1: Option[String] = Some(List(str1, str2).)

scala> res0.map(_.mkString("."))
res2: Option[String] = Some(str1.str2)
```